mirai-console/backend/real-time-test-unit/test/MiraiConsoleRealTimeTestUnit.kt

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2019-2021 Mamoe Technologies and contributors.
+ *
+ * 此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
+ * Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
+ *
+ * https://github.com/mamoe/mirai/blob/dev/LICENSE
+ */
+
+package net.mamoe.console.rttu
+
+import org.junit.jupiter.api.Test
+import java.io.File
+import java.lang.management.ManagementFactory
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlin.concurrent.thread
+
+
+class MiraiConsoleRealTimeTestUnit {
+    @Test
+    fun bootstrap() {
+        /*
+        implementation note:
+        不使用 @TempDir 是为了保存最后一次失败快照, 便于 debug
+         */
+        val workingDir = File("build/rttu") // mirai-console/backend/rttu/build/trru
+        workingDir.deleteRecursively()
```

Comment:
很危险, 建议写一个 security 文件策略如

```kotlin
        fun checkSecurityFile(workingDir: File) {
            val securityFile = workingDir.resolve(SECURITY_FILE_NAME)
            if (workingDir.exists()) {
                if (securityFile.exists()) {
                    // ok
                } else {
                    when (workingDir.listFiles()?.isEmpty()) {
                        null -> error("Working dir '${workingDir}' is not a directory. Please change `mavenCentralPublish.workingDir`")
                        false -> error("Working dir '${workingDir}' is not empty. Please change `mavenCentralPublish.workingDir`")
                        true -> {}
                    }
                }
            } else {
                // ok
            }
            workingDir.mkdirs()
            if (!workingDir.isDirectory) {
                error("Failed to create directory '${workingDir.absolutePath}'. Please change `mavenCentralPublish.workingDir`")
            }
            securityFile.writeText(
                """
                        This file is created by the Gradle plugin '$PLUGIN_ID', and this directory is reserved for publication. 
                        
                        Everytime signing artifacts, all files in this directory will be removed without notice.
                        Do not put any files into this directory, or use this directory for any other purposes, otherwise they will be deleted as stated above!
                        
                        If you are not using this directory for publication (i.e. when you changed the `workingDir`), you can delete this file.
                        Without the existence of this security file, the plugin won't remove anything and will give you an error on startup.
                        """
                    .trimIndent()
            )
        }
```

And tests:
```kotlin
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.io.File
import kotlin.test.assertFails
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class SecurityFileTest : AbstractPluginConfigurationTest() {
    @TempDir
    lateinit var tempDir: File 

    @Test
    fun `should pass if workingDir does not exists`() {
        val resolve = tempDir.resolve("empty-dir")
        assertTrue { !resolve.exists() }
        MavenCentralPublishPlugin.checkSecurityFile(resolve)
    }

    @Test
    fun `should pass if workingDir is empty`() {
        val resolve = tempDir.resolve("empty-dir")
        resolve.mkdirs()
        assertTrue { resolve.exists() && resolve.isDirectory && resolve.listFiles()!!.isEmpty() }
        MavenCentralPublishPlugin.checkSecurityFile(resolve)
    }

    @Test
    fun `should pass if workingDir is not empty with security file`() {
        val resolve = tempDir.resolve("empty-dir")
        resolve.mkdirs()
        assertTrue { resolve.exists() }
        assertTrue { resolve.isDirectory }
        tempDir.resolve(MavenCentralPublishPlugin.SECURITY_FILE_NAME).writeText("1")
        tempDir.resolve("something.txt").writeText("1")
        MavenCentralPublishPlugin.checkSecurityFile(resolve)
    }

    @Test
    fun `should fail if workingDir is file`() {
        val resolve = tempDir.resolve("empty-dir")
        resolve.writeText("1")
        assertTrue { resolve.exists() && resolve.isFile }
        assertFails { MavenCentralPublishPlugin.checkSecurityFile(resolve) }.let { exception ->
            assertTrue { exception.message!!.contains("is not a directory") }
        }
    }

    @Test
    fun `should fail if workingDir is not empty without security file`() {
        val resolve = tempDir.resolve("empty-dir")
        resolve.mkdirs()
        assertTrue { resolve.exists() && resolve.isDirectory }
        tempDir.resolve("something.txt").writeText("1")
        assertFalse { tempDir.resolve(MavenCentralPublishPlugin.SECURITY_FILE_NAME).exists() }
        MavenCentralPublishPlugin.checkSecurityFile(resolve)
    }
}
```

Reply:
最坏的情况也只是把根目录的build/rttu给删了我觉得没太大必要额外加检查